### ISSUE TITLE
Keyboard Feature for Web

### ISSUE THREAD

**Issue by AndreWohnsland** [AndreWohnsland/CocktailBerry#211]
Currently, when installing a fresh OS and CocktailBerry, there seem no out of the box virtual keyboard support. For users having touchscreens instead of mobile devices connecting (e.g. Old setup with v2) it is almost necessary to have a keyboard attached or virtual keyboard (desired state).

Suggested from [RPi bookworm feedback](https://github.com/raspberrypi/bookworm-feedback/issues/14):
```
sudo apt install wvkbd
```

## AC
- [ ] User can optional set up virtual keyboard for CocktailBerry

**Comment by AndreWohnsland** [comment]
Sqeekboard is delivered with the latest OS. It seems not to work with Chromium, but Firefox. We might use FF in Kiosk instead, since it also already bundled with Raspbian.

**Comment by AndreWohnsland** [comment]
With the new debian OS Port, it should be better supported. We should keep investigating how this will now work on fullscreen in chromium/firefox.

**Comment by AndreWohnsland** [comment]
# Virtual Keyboard on Raspberry Pi OS — State of the Art (May 2026)

## TL;DR

Out of the box, Raspberry Pi OS Bookworm/Trixie still does **not** provide a working virtual keyboard for fullscreen / kiosk Chromium or Electron applications. The default on-screen keyboard, **Squeekboard**, is installed and runs, but is rendered *beneath* fullscreen surfaces and therefore invisible the moment the browser enters kiosk or F11 fullscreen mode.
The situation is unchanged from the original investigation in summer 2025.

## Why it does not work

Raspberry Pi OS uses the **labwc** Wayland compositor (wlroots-based) and ships Squeekboard as the recommended on-screen keyboard. Wayland's `wlr-layer-shell` protocol defines a fixed stacking order of layers:

background → bottom → normal windows → always-on-top → top → FULLSCREEN → overlay → lockscreen

Squeekboard binds itself to the `top` layer.
Fullscreen surfaces sit **above** the `top` layer by protocol definition, so any fullscreen Chromium / Electron window covers the keyboard completely. Only surfaces on the `overlay` layer (or higher) would render in front of a fullscreen window — and Squeekboard does not use that layer.

This is **specification-correct behavior** of labwc; it is not a compositor bug. The constraint is hard-coded inside Squeekboard itself.

## Why it has not been fixed natively

The fix is technically trivial — a single constant change from `LAYER_TOP` to `LAYER_OVERLAY`. The reason it has not landed is organizational, not technical:

- **Upstream Squeekboard (Phosh / GNOME)** has declined to change the layer. Squeekboard is primarily a phone-shell keyboard for Phosh, where rendering above the lockscreen would be a security regression. The relevant upstream tickets ([Phosh #408](https://gitlab.gnome.org/World/Phosh/squeekboard/-/issues/408), [#438](https://gitlab.gnome.org/World/Phosh/squeekboard/-/issues/438)) have been open for years with no
  movement and an explicit "won't fix as-is" stance.

- **labwc** [closed the issue](https://github.com/labwc/labwc/issues/2926) on 2025-07-18 as not-a-labwc-bug, with the maintainer pointing out that the change belongs in Squeekboard.

- **Raspberry Pi's downstream Squeekboard fork** ([raspberrypi-ui/squeekboard](https://github.com/raspberrypi-ui/squeekboard)) could carry a Debian patch independently of upstream. A Raspberry Pi engineer acknowledged the problem in July 2025 and said he would discuss it internally. No layer-related commit has landed in the repository since (latest non-trivial commit: 2025-03-12). The tracking [issue #13](https://github.com/raspberrypi-ui/squeekboard/issues/13) remains open, with the last activity 2026-01-15 ("Is that env var fix possible today?" — unanswered).

So the gap is well understood by every party involved; no single maintainer has taken ownership of shipping a fix.

## What users currently see

| Scenario                              | Behavior                                                        |
| ------------------------------------- | --------------------------------------------------------------- |
| Chromium / Firefox in a normal window | Squeekboard appears on focus (mostly works)                     |
| Chromium `--kiosk`                    | Squeekboard runs but is invisible                               |
| Chromium `--start-fullscreen` / F11   | Squeekboard runs but is invisible                               |
| Electron app in fullscreen            | Squeekboard runs but is invisible                               |
| Single-display setup                  | Often appears not to launch at all (it does — it's just hidden) |

The behavior is identical whether the OS image is Bookworm or the newer Trixie port. Switching between Wayfire and labwc does not change the outcome; both respect the same layer-shell ordering.

## Outlook

A native, out-of-the-box fix in the near term looks **unlikely**:

- Upstream Phosh has a principled reason not to change the default layer (lockscreen security). Any upstream solution would need to be a new configuration knob — an env var, a GSetting, or a D-Bus method — which means a design discussion, not a one-line patch. There is no indication that such a discussion is active.
- The Raspberry Pi downstream fork is the more realistic vehicle for a fix, but has shown no commits on this topic in roughly 14 months despite ongoing user reports. The kiosk / fullscreen use case is a niche relative to Raspberry Pi OS's primary desktop audience, which reduces priority.
- The Trixie-based Raspberry Pi OS release does not change anything structurally; it ships the same Squeekboard fork with the same downstream patch set.
- The broader Wayland ecosystem (wlroots, KDE, GNOME) considers the current layer ordering correct, so no protocol-level change is on the table either.

A realistic best case is that a Raspberry Pi engineer eventually applies a Debian patch adding an env-var or GSetting opt-in, similar to the community proposals. A realistic worst case is that nothing changes and kiosk-style projects continue to depend on third-party workarounds indefinitely.

## References

- [labwc #2926 — Squeekboard not visible over fullscreen Chromium](https://github.com/labwc/labwc/issues/2926)
- [raspberrypi-ui/squeekboard #13 — Squeekboard does not display when in fullscreen](https://github.com/raspberrypi-ui/squeekboard/issues/13)
- [Phosh squeekboard #408 — layer-shell / lockscreen concern](https://gitlab.gnome.org/World/Phosh/squeekboard/-/issues/408)
- [Phosh squeekboard #438 — related upstream discussion](https://gitlab.gnome.org/World/Phosh/squeekboard/-/issues/438)
- [Raspberry Pi forum — Onscreen Keyboard doesn't work with Chromium in kiosk/F11](https://forums.raspberrypi.com/viewtopic.php?t=390053)
- [Raspberry Pi forum — HOW-TO Bookworm On-screen Keyboard](https://forums.raspberrypi.com/viewtopic.php?t=358147)
- [bookworm-feedback #14 — No on-screen keyboard available in Wayland](https://github.com/raspberrypi/bookworm-feedback/issues/14)

**Comment by AndreWohnsland** [comment]
we will implement an optional feature using https://github.com/hodgef/react-simple-keyboard (opt in) for touchscreen setups to activate in V2, being able to run in fullscreen.
